src/qz/printer/info/CachedPrintService.java

```diff
@@ -0,0 +1,130 @@
+package qz.printer.info;
+
+import qz.common.CachedObject;
+
+import javax.print.*;
+import javax.print.attribute.Attribute;
+import javax.print.attribute.AttributeSet;
+import javax.print.attribute.PrintServiceAttribute;
+import javax.print.attribute.PrintServiceAttributeSet;
+import javax.print.event.PrintServiceAttributeListener;
+import java.util.HashMap;
+import java.util.function.Supplier;
+
+public class CachedPrintService implements PrintService {
+    public PrintService innerPrintService;
+    // PrintService.getName() is slow, use a cache instead per JDK-7001133
+    // TODO: Remove this comment when upstream bug report is filed
+    private static final long lifespan = CachedObject.DEFAULT_LIFESPAN;
+    private static final CachedObject<PrintService> cachedDefault = new CachedObject<>(CachedPrintService::innerLookupDefaultPrintService, lifespan);
+    private static final CachedObject<PrintService[]> cachedPrintServices = new CachedObject<>(CachedPrintService::innerLookupPrintServices, lifespan);
+    private final CachedObject<String> cachedName;
+    private final CachedObject<PrintServiceAttributeSet> cachedAttributeSet;
+    private final HashMap<Class<?>, CachedObject<?>> cachedAttributes = new HashMap<>();
+
+    public static PrintService lookupDefaultPrintService() {
+        return cachedDefault.get();
+    }
+
+    public static PrintService[] lookupPrintServices() {
+        return cachedPrintServices.get();
+    }
+
+    private static PrintService innerLookupDefaultPrintService() {
+        return new CachedPrintService(PrintServiceLookup.lookupDefaultPrintService());
+    }
+
+    private static PrintService[] innerLookupPrintServices() {
+        PrintService[] printServices = PrintServiceLookup.lookupPrintServices(null, null);
+        for (int i = 0; i < printServices.length; i++) {
+            printServices[i] = new CachedPrintService(printServices[i]);
+        }
+        return printServices;
+    }
```

Comment:
Do these static functions belong here?  They seem like they're (mostly) unrelated to the class' function and may be better suited for a dedicated class, e.g. `PrintServiceLookupCache`, etc.

Reply:
yea they could exist in a dedicated class. They could even live in printServiceMatcher. The only reason they are here is because I wanted all of the caching stuff in one place